File: src/lib/scala/Slate.Common/src/main/scala/slate/common/templates/TemplateParser.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.common.templates

import slate.common._
import slate.common.lex.LexState
import slate.common.results.ResultCode
import slate.common.results.ResultFuncs._

import TemplateConstants._

/**
  * Parses a text template that can contain variables/substitutions.
  * e.g. An email template such as :
  *
  * "Hi @{user.name}, Welcome to @{startup.name}, please click @{verifyUrl} to verify your email."
  *
  * Is parsed into individual Template parts
  * 1. "Hi"                    -> text
  * 2. "user.name"             -> substitution
  * 3. ", Welcome to "         -> text
  * 4. "startup.name"          -> substitution
  * 5. ", please click"        -> text
  * 6. "verifyUrl"             -> substitution
  * 7. " to verify your email" -> text
  *
  * @param text
  */
class TemplateParser(val text:String) {

  class ParseState extends LexState(text, Some(text.length - 1)) {
  }


  private val _state = new ParseState()


  def parse(): Result[List[TemplatePart]] = {

    val subs = new scala.collection.mutable.ListBuffer[TemplatePart]
    val result =
      try {
        var lastText: String = ""
        var lastPos = _state.pos

        while (_state.pos <= _state.END) {
          val c = _state.text(_state.pos)
          val hasMore = _state.pos < _state.END
          val n = if (hasMore) _state.text(_state.pos + 1) else ' '

          _state.count = _state.count + 1

          // CASE 1: substitution
          if (c == '@' && hasMore && n == '{') {
            if (!Strings.isNullOrEmpty(lastText)) {
              val sub = new TemplatePart(lastText, TypeText, lastPos, _state.pos - lastPos)
              subs.append(sub)
            }
            val sub = readSub()
            subs.append(sub)

            // reset
            lastText = ""
            lastPos = _state.pos
          }
          // CASE 2: Keep reading until sub
          else {
            lastText += c
            _state.pos += 1
          }
        }

        // Last part was text ?!
        if (!Strings.isNullOrEmpty(lastText)) {
          val sub = new TemplatePart(lastText, TemplateConstants.TypeText, lastPos, _state.pos - lastPos)
          subs.append(sub)
        }

        // Success!
        success(subs.toList)
      }
      catch
        {
          case ex:Exception =>
          {
            val line = _state.line
            val text = s"Error occurred at : line : $line, char : ${_state.pos}" + ex.getMessage
            val err = new Exception( text, ex )
            failure(msg = Some(text), err = Some(err))
          }
        }
    result
  }


  /**
    * reads a substitution inside of @{}
    * expects that pos/current char = @ and next char = {
 *
    * @return
    */
  def readSub(): TemplatePart = {

    advanceAndExpect('{')
    val c = advance()
    val start = _state.pos
    val end = start
    // 1. edge case ${}
    if( c == '}') {
      advance()
      new TemplatePart("", 0, start, end)
    }
    else {
      // 2. read sub
      Loops.doUntil({
        if(_state.pos <= _state.END) {
          val curr = _state.text(_state.pos)
          if (curr == '}') {
            false
          }
          else if ((_state.pos + 1 <= _state.END)) {
            _state.pos += 1
            true
          }
          else {
            false
          }
        }
        else {
          false
        }
      })

      val text = _state.substringInclusive(start)
      val sub = new TemplatePart(text, TypeSub, start, _state.pos)
      _state.pos += 1
      sub
    }
  }


  def advanceAndExpect(expected:Char): Char = {
    val ch = advance()
    require(ch == expected, s"Expected ${expected} but found ${ch}")
    ch
  }


  def advance(): Char = {
    _state.pos += 1
    _state.text(_state.pos)
  }
}

File: src/common/utils/src/main/kotlin/kiit/utils/templates/TemplateParser.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package kiit.utils.templates

import kiit.common.lex.LexState
import kiit.common.utils.Loops
import kiit.results.Failure
import kiit.results.Success
import kiit.results.Try

/**
 * Parses a text template that can contain variables/substitutions.
 * e.g. An email template such as :
 *
 * "Hi @{user.name}, Welcome to @{startup.name}, please click @{verifyUrl} to verify your email."
 *
 * Is parsed into individual Template parts
 * 1. "Hi"                    -> text
 * 2. "user.name"             -> substitution
 * 3. ", Welcome to "         -> text
 * 4. "startup.name"          -> substitution
 * 5. ", please click"        -> text
 * 6. "verifyUrl"             -> substitution
 * 7. " to verify your email" -> text
 *
 * @param text
 */
class TemplateParser(val text: String) {

    private val state = LexState(text)

    fun parse(): Try<List<TemplatePart>> {

        val subs = mutableListOf<TemplatePart>()
        val result =
                try {
                    var lastText: String = ""
                    var lastPos = state.pos

                    while (state.pos < state.END) {
                        val c = state.text[state.pos]
                        val hasMore = state.pos + 1 < state.END
                        val n = if (hasMore) state.text[state.pos + 1] else ' '

                        state.count = state.count + 1

                        // CASE 1: substitution
                        if (c == '@' && hasMore && n == '{') {
                            if (!lastText.isNullOrEmpty()) {
                                val sub = TemplatePart(lastText, TemplateConstants.TypeText, lastPos, state.pos - lastPos)
                                subs.add(sub)
                            }
                            val sub = readSub()
                            subs.add(sub)

                            // reset
                            lastText = ""
                            lastPos = state.pos
                        }
                        // CASE 2: Keep reading until sub
                        else {
                            lastText += c
                            state.pos += 1
                        }
                    }

                    // Last part was text ?!
                    if (!lastText.isNullOrEmpty()) {
                        val sub = TemplatePart(lastText, TemplateConstants.TypeText, lastPos, state.pos - lastPos)
                        subs.add(sub)
                    }

                    // Success!
                    Success(subs.toList())
                } catch (ex: Exception) {
                    val line = state.line
                    val text = "Error occurred at : line : $line, char : ${state.pos}" + ex.message
                    val err = Exception(text, ex)
                    Failure(err, msg = text)
                }
        return result
    }

    /**
     * reads a substitution inside of @{}
     * expects that pos/current char = @ and next char = {
     *
     * @return
     */
    fun readSub(): TemplatePart {

        advanceAndExpect('{')
        val c = advance()
        val start = state.pos
        val end = start
        // 1. edge case ${}
        return if (c == '}') {
            advance()
            TemplatePart("", 0, start, end)
        } else {
            // 2. read sub
            Loops.doUntil {
                if (state.pos < state.END) {
                    val curr = state.text[state.pos]
                    if (curr == '}') {
                        false
                    } else if ((state.pos + 1 < state.END)) {
                        state.pos += 1
                        true
                    } else {
                        false
                    }
                } else {
                    false
                }
            }

            val text = state.substringInclusive(start)
            val sub = TemplatePart(text, TemplateConstants.TypeSub, start, state.pos)
            state.pos += 1
            sub
        }
    }

    fun advanceAndExpect(expected: Char): Char {
        val ch = advance()
        require(ch == expected) { "Expected $expected but found $ch" }
        return ch
    }

    fun advance(): Char {
        state.pos += 1
        return state.text[state.pos]
    }
}
